reasoner/resolution/framework/ResponseProducer.java

```diff
@@ -29,17 +31,17 @@
 
 public class ResponseProducer {
     private final Set<ConceptMap> produced;
-    private final Iterator<ConceptMap> traversalProducer;
+    private final ResourceIterator<AnswerState.UpstreamVars.Derived> traversalProducer;
     private final LinkedHashSet<Request> downstreamProducer;
     private final int iteration;
     private Iterator<Request> downstreamProducerSelector;
 
-    public ResponseProducer(Iterator<ConceptMap> traversalProducer, int iteration) {
-        this(traversalProducer, iteration, new HashSet<>());
+    public ResponseProducer(ResourceIterator<AnswerState.UpstreamVars.Derived> upstreamAnswers, int iteration) {
+        this(upstreamAnswers, iteration, new HashSet<>());
     }
 
-    private ResponseProducer(Iterator<ConceptMap> traversalProducer, int iteration, Set<ConceptMap> produced) {
-        this.traversalProducer = traversalProducer;
+    private ResponseProducer(ResourceIterator<AnswerState.UpstreamVars.Derived> upstreamAnswers, int iteration, Set<ConceptMap> produced) {
+        this.traversalProducer = upstreamAnswers.filter(derived -> !hasProduced(derived.withInitialFiltered()));
```

Comment:
immediately filter answers that are going to be returned according to the `produced` map